instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/rules/MockServerActivityTestRule.kt

```diff
@@ -0,0 +1,140 @@
+import androidx.test.rule.ActivityTestRule
+import com.datadog.android.Datadog
+import com.datadog.android.log.EndpointUpdateStrategy
+import com.google.gson.JsonElement
+import com.google.gson.JsonParser
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.util.zip.GZIPInputStream
+import okhttp3.Headers
+import okhttp3.mockwebserver.Dispatcher
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.RecordedRequest
+
+internal class MockServerActivityTestRule<T : Activity>(
+    activityClass: Class<T>,
+    val keepRequests: Boolean = false
+) : ActivityTestRule<T>(activityClass) {
+
+    data class HandledRequest(
+        val headers: Headers? = null,
+        val method: String? = null,
+        val jsonBody: JsonElement? = null,
+        val testBody: String? = null
+    )
+
+    private val mockWebServer: MockWebServer = MockWebServer()
+
+    private val requests = mutableListOf<HandledRequest>()
+
+    // region ActivityTestRule
+
+    override fun beforeActivityLaunched() {
+        requests.clear()
+        mockWebServer.start()
+        mockWebServer.setDispatcher(object : Dispatcher() {
+            override fun dispatch(request: RecordedRequest): MockResponse {
+                if (keepRequests) {
+                    handleRequest(request)
+                } else {
+                    Log.w(TAG, "Dropping @request:$request")
+                }
+
+                return mockResponse(200)
+            }
+        })
+
+        val fakeEndpoint = mockWebServer.url("/").toString().removeSuffix("/")
+        Datadog.setEndpointUrl(fakeEndpoint, EndpointUpdateStrategy.DISCARD_OLD_LOGS)
+        super.beforeActivityLaunched()
+    }
+
+    override fun afterActivityFinished() {
+        mockWebServer.shutdown()
+        InstrumentationRegistry.getInstrumentation().context.filesDir?.deleteRecursively()
+        super.afterActivityFinished()
+    }
+
+    // endregion
+
+    // region MockServerRule
+
+    fun getRequests(): List<HandledRequest> {
+        Log.i(TAG, "Caught ${requests.size} requests")
+        return requests.toList()
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun mockResponse(code: Int): MockResponse {
+        return MockResponse()
+            .setResponseCode(code)
+            .setBody("{}")
+    }
+
+    private fun handleRequest(request: RecordedRequest) {
+        Log.i(TAG, "Handling @request:$request")
+        val encoding = request.headers.get(HEADER_CONTENT_ENCODING)
+        val contentType = request.headers.get(HEADER_CONTENT_TYPE)
+        val content = if (encoding == "gzip") {
+            request.unzip()
+        } else {
+            String(request.body.readByteArray(), Charsets.UTF_8)
+        }
+
+        val jsonBody = if (contentType == "application/json") {
+            JsonParser.parseString(content)
+        } else {
+            null
```

Comment:
You don't need to add a "text/plain" case : the HandledRequest object has a `JsonElement` field when the data is Json, and always has a `textBody` which is the plain text content. When dealing with plain text request, the `jsonBody` will be null, but the `textBody` will the body content as plain text.

Reply:
Yeah...but still I need to parse this body as a JsonArray...this is my intention, otherwise you cannot properly assert the sent data. For me each HandledRequest should always have jsonBody (no matter in which format the data was sent) and in your tests you should not care about how the data was stored in the end you should get an array of Json objects to assert.